### ISSUE TITLE
[MP][Scenarios][Gameplay] Add more unique content scenarios.

### ISSUE THREAD

**Issue by knyghtmare** [wesnoth/wesnoth#7365]
### Describe the desired feature

We have currently:
- 24: (pvp deathmatch scenarios)
- 4 survivals (DF/IoM/ANL/TS)
- 6 (3p/5p) FFA
- 24 mix of team deathmatch and/or FFA.

So, categorizing, it's only 3 types of MP content in core package.
That's not good. It's stale. 
I suggest more unique content.

### Deliverables:

- [ ] 4p RPG scenario over big map (one epic quest in it + side quests)
- [ ] Quest-type scenarios (Not just the Wilderlands reworked one, another one with it)
- [ ] Domination Game mode
- [ ] CTF game mode

### Context for Demand
1. New generation players expect something more than the common game types. 
2. They expect to have a variety of game modes but are met with the 3 generic types...
3. Hence, new players are demotivated.
4. Giving "there are add-ons" isn't really a strong argument to not add them

**Comment by Pentarctagon** [comment]
CTF meaning Capture The Flag?

**Comment by knyghtmare** [comment]
> CTF meaning Capture The Flag?

yes, updated post with clarification

**Comment by soliton-** [comment]
I would not add content to check boxes. Add content because it is good. So if there is good/popular UMC then consider adding it to mainline especially if the original author would like to maintain it as part of mainline.

Just adding more content to mainline that needs to be maintained by someone or worse constantly adding and removing content because otherwise it gets stale does not sound good.

**Comment by CelticMinstrel** [comment]
Turn King of the Hill into an actual [king of the hill](https://ageofempires.fandom.com/wiki/King_of_the_Hill) scenario. That is to say, the victory condition is to control the central keep for a specified number of turns.

**Comment by Hejnewar** [comment]
> Capture the Flag (CTF) game mode

Actually very hard to do. I do not recomend.

> Domination Game mode

Explain, is this something that I work on or something else?

**Comment by CelticMinstrel** [comment]
> Actually very hard to do.

Why do you think it's hard to do?

**Comment by Hejnewar** [comment]
Because I think in category of fun and difference to the standard gameplay, not code. It is easy to code, but it is very hard to do it well enough.

**Comment by CelticMinstrel** [comment]
That doesn't really answer the question though. Why do you think it's hard to do it "well enough"?

**Comment by stevecotton** [comment]
> 4p RPG scenario over big map (one epic quest in it + side quests)

I'm opposed to putting any multiplayer RPG into mainline, at least until we either:
* resolve the problems with patching World Conquest in the stable branch, or
* have a (circa 2 year) stable release cycle when we don't find any major bugs in World Conquest.

**Comment by knyghtmare** [comment]
> That doesn't really answer the question though. Why do you think it's hard to do it "well enough"?

Because the challenge is hyper tuning map, mechanics, penalties, anomalies and others to allow fun and optimized gameplay without dragging on too long.

https://github.com/knyghtmare/knyght_CTF

> Explain, is this something that I work on or something else?

In Domination, 2-4 teams try to dominate the map by capturing and keeping control points on the map and hampering capture/holding efforts of the enemy teams. Each control point that a team has captured will generate points over time. Victory is given if points surpass some threshold or turns out (highest point holding team(s) are declared winners).

> I'm opposed to putting any multiplayer RPG into mainline, at least until we either:

How is World Conquest related to role-playing? :confused: I do suppose WC needs some attention on bug patching.

**Comment by CelticMinstrel** [comment]
> How is World Conquest related to role-playing?

Probably that the level of complexity is similar?

**Comment by knyghtmare** [comment]
> Ah, so the goal is to capture and hold all the villages.

Not sure how you connected a village to a control point...I mean control points. Areas of flat terrain that have to be defended. Having village as control point seems not fair as some factions have high defence units on villages while some have low. Flat negates that.

**Comment by Pentarctagon** [comment]
> How is World Conquest related to role-playing? confused I do suppose WC needs some attention on bug patching.

Because the benefit of the current set of available scenarios is that most of them are very simple - some with almost no scenario code at all. What WC has brought attention to is that not being able to fix significant bugs that affect gameplay in the stable branch for 1-2 years because we can't break backwards compatibility with earlier stable releases is a very negative thing. So adding more scenarios that are similarly complex would make that problem even worse and more obvious to players without first resolving #7288.

**Comment by CelticMinstrel** [comment]
> Not sure how you connected a village to a control point...I mean control points. Areas of flat terrain that have to be defended. Having village as control point seems not fair as some factions have high defence units on villages while some have low. Flat negates that.

Yeah, I get your point, but villages already provide the exact mechanic required, so why not use it? You could just define a terrain that gives income but aliases to flat. (But then you wouldn't use regular villages, of course. You could define terrains similar to the oasis that give healing but not income.) And then you can easily see who currently has control of the point, too, because a flag is automatically added. It's way less coding work to simple leverage the existing village mechanic, isn't it?

**Comment by Hejnewar** [comment]
I wont say anything about RPG or Quest right now, but neither CTF nor Domination pass my quality standard. The reason is simple, with unchanged main game mechanics mainly income, upkeep, time of day (and in possibly also zoc) players simply wont play objective, the game doesnt change from simple unit stopm with an added timer (or in case of CTF drawing out the game over a longer period of time, because you have to carry all these flags after winning). 

Yes I did recently start to theorise the domination mode, but I straight up abanddoned tod and standard economy model (and zoc) and it would work well with special era only anyway. In order to make it playable with anything (or most eras existing) domination is bad althogether and should be replaced with DTO (Destroy the Objectives) mode as it has way more pros, such as correct unit capping power reflection (which is the most important here if you dont want to just see fast units everywhere), better as well as more natural better player punishment (which also helps with luck).

The goal should be to create, faster (around max 24 turns avg. still 16 turns), more noob friendly (aka slight more punishing the winning player), engaging (aka it is not over till the end), modes, that are quiet a bit different form existing gameplay. And remember player will play to win not to play the way you want to play the map.  

>So adding more scenarios that are similarly complex would make that problem even worse and more obvious to players without first resolving https://github.com/wesnoth/wesnoth/issues/7288.

WC is overly complex, I dont see wwhy quest scenarios would need to be this complex, you can make good stuff with simple code. (I know cuz I suck at advanced coding.)

**Comment by CelticMinstrel** [comment]
> neither CTF nor Domination pass my quality standard

I don't think that's a fair statement at all. The quality of a CTF or Domination scenario could vary wildly; to judge them just based on the label before you've even seen them is unfair.

> with unchanged main game mechanics mainly income, upkeep, time of day (and in possibly also zoc)

A custom scenario can do whatever it wants with the game mechanics. You could set upkeep to 0, disable standard recruiting, and use income as a scorekeeper, for example.

> should be replaced with DTO (Destroy the Objectives) mode

Ah yes, this sounds like another interesting game mode. Whoever plunders and destroys the most villages wins.

**Comment by Hejnewar** [comment]
> I don't think that's a fair statement at all. The quality of a CTF or Domination scenario could vary wildly; to judge them just based on the label before you've even seen them is unfair.

Well thats not really an issue since I did see them. Also its kinda my job to be able to judge stuff without seeing it / a lot of it.

**Comment by CelticMinstrel** [comment]
> Well thats not really an issue since I did see them.

Then you can say none of the existing ones pass your quality standard. That's not the same as saying CTF or domination could never pass your standard.

**Comment by Hejnewar** [comment]
> Then you can say none of the existing ones pass your quality standard. That's not the same as saying CTF or domination could never pass your standard.

Thats fair I will rephrase it. I lack imagination to imagine a scenario in CTF mode that would pass my quality standard. I can possibly see a Domination scenario that maybe would be fine but I still see DTO as superior option with similar characteristics. (In case of somewhat standard 1v1 / 2v2 / etc.) 

> Ah yes, this sounds like another interesting game mode. Whoever plunders and destroys the most villages wins.

Villages are actually bad, if they work like villages, I would probably prefer to have stationary units in certain places of the map, that have quiet high hp (in hundreds) with some weak attack in both ranges.

**Comment by CelticMinstrel** [comment]
> Villages are actually bad, if they work like villages

Well, if the goal is to "destroy" the villages then I guess that means they wouldn't work like villages, because they'd turn to rubble when you step on them, or something. But using stationary units also sounds like an interesting idea.

**Comment by Jonathan-Kelly** [comment]
> 4p RPG scenario over big map (one epic quest in it + side quests)

There are quite a few of these already made by Bob_The_Mighty (who also created mainline's _A New Land_ scenario).

> Quest-type scenarios (Not just the Wilderlands reworked one, another one with it)

I can surely look at doing a second one, though I would want to get Hejnewar's feedback and approval on [The Wilderlands Quest](https://forums.wesnoth.org/viewtopic.php?t=56216) first.

> Turn King of the Hill into an actual [king of the hill](https://ageofempires.fandom.com/wiki/King_of_the_Hill) scenario. That is to say, the victory condition is to control the central keep for a specified number of turns, or something along those lines.

[Essentially done.](https://forums.wesnoth.org/viewtopic.php?t=56209&sid=70b7bb1e5a1bba35dcbc34faf10a60b3)

I am also thinking about trying something similar (but lighter) for Forest of Fear. Have small numbers of dangerous wildlife randomly wander in from the edges of the map to attack player units and each other, but then become allies to any player who holds the central keep (for as long as it is held).

**Comment by Hejnewar** [comment]
> I can surely look at doing a second one, though I would want to get Hejnewar's feedback and approval on [The Wilderlands Quest](https://forums.wesnoth.org/viewtopic.php?t=56216) first.

I will try to do this this week as I have 2 free days, but no promises since I have some other stuff planned already.

**Comment by CelticMinstrel** [comment]
> Essentially done.

It sounds interesting, but from your description it doesn't sound like it's quite what I had in mind. I was suggesting the victory condition should be to stay in control of the hill for a specified number of turns. Though I guess if only one side is standing for any reason, they should also be the winners, but if one side captures the hill and holds it for (say) 10 turns, that side should win even if there are still other sides standing. (10 is just a random number, I don't know what number would actually make sense; in fact I'd probably suggest making it configurable.)

**Comment by Discontinuum** [comment]
To be honest, I wonder if, let say, World Conquest would be better in the form of the addon supported by the core team. Adding to mainline caused those great troubles with being unable to update it. An addon though can be updated easily. It could've been an "essential addon" maybe, one that is installed when you install the game.

With the current state in updating matters, the mp mainline content would better be either really simple or fully inlineable (so the update won't cause any oos because everything is provided the host of the game anyway)

**Comment by Jonathan-Kelly** [comment]
Thanks, Hejnewar. I uploaded its latest version to the 1.16 add-ons server as _"Improved The Wilderlands"_. When you get the chance, _"Improved King of the Hill"_ is there also and ready for your review.

> WC is overly complex, I dont see why quest scenarios would need to be this complex, you can make good stuff with simple code.

Hmm, do you have an idea of which features you might want to see removed from World Conquest to streamline its game play and simplify its code?

> I was suggesting the victory condition should be to stay in control of the hill for a specified number of turns.

That will accomplish about the same thing but with some interesting differences related to Hejnewar's definition of an _engaging_ scenario. Your suggestion has the advantage that the losing sides may still have some useful recruiting power until 3 or 4 turns before the end. The disadvantage is the losing sides will not be able to equalize the situation by capturing the winning side's villages, as time will still be against them regardless.

A third option could be that taking the hill halves `village_gold` for all enemy sides. So they can still recruit, but are at an income disadvantage which will become decisive before not too long.

**Comment by Hejnewar** [comment]
> Thanks, Hejnewar. I uploaded its latest version to the 1.16 add-ons server as "Improved The Wilderlands". When you get the chance, "Improved King of the Hill" is there also and ready for your review.

Okay.

> Hmm, do you have an idea of which features you might want to see removed from World Conquest to streamline its game play and simplify its code?

I meant it the other way aroud, you dont need the complexity of WC to create something good and entertaining. 

> That will accomplish about the same thing but with some interesting differences related to Hejnewar's definition of an engaging scenario.

Opinion that I shared is mostly relevant when it comes to somewhat standard 1v1, other more different modes probably should have them slightly different anyway.

**Comment by sevu** [comment]
That one can update WC2 (and also Dark Forecast) on 1.16 only to some extend is due to the way it's implemented and could also be done differently. A new game mode can be as complex as you want.

**Comment by CelticMinstrel** [comment]
Another possible game mode could be "collect them all". In Age of Empires II there is a mode where you must gather all the relics scattered across the map and bring them back to your base in order to win. It wouldn't be too hard to code something similar in Wesnoth. (It could probably be done as a `[modification]`, actually, though I'm not sure if that's better or worse than having a dedicated map.)

**Comment by doofus-01** [comment]
> To be honest, I wonder if, let say, World Conquest would be better in the form of the addon supported by the core team. Adding to mainline caused those great troubles with being unable to update it. An addon though can be updated easily. It could've been an "essential addon" maybe, one that is installed when you install the game.

This general "official add-on" idea could also address all the content people want to merge into core (e.g. #7335, #5532).  If it's used as a development arena, rather than a second-tier dumping ground, it might even be a simple way to get more assistance or contributions from the player base.

**Comment by knyghtmare** [comment]
> This general "official add-on" idea could also address all the content people want to merge into core (e.g. https://github.com/wesnoth/wesnoth/issues/7335, https://github.com/wesnoth/wesnoth/pull/5532). If it's used as a development arena, rather than a second-tier dumping ground, it might even be a simple way to get more assistance or contributions from the player base.

I suppose it would lighten the game a bit if some WIP pull requests of unit additions got in that way, and open up community additions as some suggestions do not meet standards that much.

**Comment by Pentarctagon** [comment]
To be clear, my proposal in #7288 for "official add-on(s)" is meant as a way to make it easier to provide updates and improvements to content that's already in mainline currently. I didn't mean it as a less official or less supported way to add content that's not consistent with the rest of mainline or doesn't meeting mainline's quality standard - an official add-on would be subject to the same standards as everything else that's proposed to mainline and the expectation is it that updating it between versions would likewise be the responsibility of the development team.

If people want a sort of staging area for content that's not quite ready for mainline but might eventually be ready then that's something that could be discussed, but that would be something separate.

**Comment by Discontinuum** [comment]
> could also address all the content people want to merge into core 

I haven't checked the exact units proposed there but I only notice that adding more units to the core helps making MP addons that don't require other players to download them, so I rather like the expansion of the unit core. That doesn't mean to include everything without any selection ofc

**Comment by Hejnewar** [comment]
Yeah I really dont think this issue is a great place for "official add-on(s)" discussion, but FTR, i dont think making add-on cathegories for candidates is a good idea. 

> Another possible game mode could be "collect them all". In Age of Empires II there is a mode where you must gather all the relics scattered across the map and bring them back to your base in order to win.

Yes this could possibly work but map / scenario specifically tailored for it is required. Most likely it can't have standard economy / recruitement method. Note that collect them all can also be kill them all or similar.

**Comment by doofus-01** [comment]
> To be clear, my proposal in https://github.com/wesnoth/wesnoth/issues/7288 for "official add-on(s)" is meant as a way to make it easier to provide updates and improvements to content that's already in mainline currently. I didn't mean it as a less official or less supported way to add content that's not consistent with the rest of mainline or doesn't meeting mainline's quality standard - an official add-on would be subject to the same standards as everything else that's proposed to mainline and the expectation is it that updating it between versions would likewise be the responsibility of the development team.

If the "official add-on" mechanism is implemented, why can't there be both?  Some people will not like having Wesnoth automatically reach out to and download data from headquarters, so there will have to be some preferences option and this would just end up being a second radio button.

**Comment by Pentarctagon** [comment]
The idea is that it would be something they select manually on first start from a dialog, then update it afterwards the same way as any other add-on.

**Comment by Jonathan-Kelly** [comment]
> > Another possible game mode could be "collect them all". In Age of Empires II there is a mode where you must gather all the relics scattered across the map and bring them back to your base in order to win.
> 
> Yes this could possibly work but map / scenario specifically tailored for it is required. Most likely it can't have standard economy / recruitement method. Note that collect them all can also be kill them all or similar.

The Wilderlands map may be naturally tailored for this, with its many distinctive regions, each of which could hold a thematically fitting relic. Perhaps there should be an odd number of relics and you only need carry off a simple majority of them, so that there is no need to hunt down every player who has just one relic in their possession.

@Hejnewar for a nonstandard economy what if the scenario dropped village gold from 2 to 1 and village support from 1 to 0; do you think that would be enough?

Or perhaps turn village gold down to 0 and keep village support at 1? Then players would receive only a trickle of capped income for new recruits, and only if they held enough villages to support their existing recruits.

**Comment by knyghtmare** [comment]
> The Wilderlands map may be naturally tailored for this.

The map's already being used for it's own scenario. Maybe make a new one 50x50 or larger map? I think we have enough enough new terrains in core now to warrant a new map, perhaps with elevated terrain and the seaweed/kelp, and other new stuff. We added so many new terrain and we hardly have showcased them.

**Comment by Hejnewar** [comment]
Actually, this whole convo gave me an idea, why not just adapt For the Lord system for Wilderlands? 

The system works by giving you certain amount of points for doing certain actions, originally it was destroying (to put it lightly) a city with a bunch of goblins. Here it would be killing monsters, killing bosses, doing quests like destroy nest of X, grants even more points etc. 

I would also greatly reduce the number of villages and I would make the, give only support. All the gold would come form... Also killing monsters. I dont have a clear idea about PvP interactions for this tho. It is something to think about.

**Comment by Jonathan-Kelly** [comment]
> Here it would be killing monsters, killing bosses, doing quests like destroy nest of X, grants even more points etc.

> All the gold would come form... Also killing monsters.

Alright, I can throw together some hunting mechanics. How do you want the gold reward for killing a monster to be calculated? Like 10% of the creature's cost? Or 50% ?

We could also make the gold itself be the points? _"Get 1000 gold to win."_ So players have to decide how much to spend recruiting additional hunters and bodyguards versus how much to bank towards their victory score.

> The map's already being used for it's own scenario. Maybe make a new one 50x50 or larger map?

I think that once we have the new scenario concepts working well, creating some new maps to fit their requirements will be a lot easier. And some maps might support more than one scenario / game type.

**Comment by Hejnewar** [comment]
> Alright, I can throw together some hunting mechanics. How do you want the gold reward for killing a monster to be calculated? Like 10% of the creature's cost? Or 50% ?

I would say 10% but thats to be tested. 

> We could also make the gold itself be the points? "Get 1000 gold to win." So players have to decide how much to spend recruiting additional hunters and bodyguards versus how much to bank towards their victory score.

Well thats possible, I need to think about what would be better but having more liberty with gold is not bad I guess.
